### DISCUSSION TITLE
Run multiple commands in a custom_target

### DISCUSSION THREAD

**Discussion by ghost** [mesonbuild/meson#11869] [Q&A]
Hi,

I am currently converting my C kernel to the meson build system, and I'm having a bit of a hickup generating an iso file from an elf file;

I've successfully written the part where I copy every files I need into its own directory to then generate an ISO, the only problem that remains now is : is there a possibility to make a custom_target depend on another one?
I'm asking this question because it is required to create the directory that will contain every file needed to create the iso, and it is also required that all of the files are correctly moved into that previous directory to then run xorriso (which is the tool I use to create the ISO).

This isa part of my current buildscript
```
kernel = executable('kernel.elf',
    src,
    include_directories : inc,
    link_args : [ '-Wl,-T' + ldscript ],
    link_depends : [ ldscript ],
)

cp = find_program('cp')
xorriso = find_program('xorriso')



kernel_iso = custom_target('limine_setup',
  input : ['limine.cfg', 'external/limine/bin/limine.sys', 'external/limine/bin/limine-cd.bin', 'external/limine/bin/limine-cd-efi.bin'],
  output : ['kernel.iso'],
  command : [cp, '@INPUT@', 'iso_root'])
```

**Comment by eli-schwartz** [comment]
> is there a possibility to make a custom_target depend on another one?

Yes, simply list the assigned variable e.g. `foobar = custom_target(...)` where foobar is then the *input* to another custom target.

Probably the best way to do this is to have your `cp` custom_target inside of iso_root/meson.build, and specify that the second parameter to cp is `'@OUTDIR@'`.

Note that a single custom_target can't have multiple commands in shell syntax, because the `command:` kwarg to custom_target doesn't accept shell syntax. You could of course create a shell script or python script that internally does multiple things, although I don't think you need to -- like I said, if you put the custom_target in the subdir meson.build then you can tell meson to output it to the build directory's `iso_root/` directory and the directory will be created for you, no need to e.g. chain together `mkdir ... && cp ...`

**Reply by ghost** [reply]
> Yes, simply list the assigned variable e.g. `foobar = custom_target(...)` where foobar is then the _input_ to another custom target.
> 
> Probably the best way to do this is to have your `cp` custom_target inside of iso_root/meson.build, and specify that the second parameter to cp is `'@OUTDIR@'`.
> 
> Note that a single custom_target can't have multiple commands in shell syntax, because the `command:` kwarg to custom_target doesn't accept shell syntax. You could of course create a shell script or python script that internally does multiple things, although I don't think you need to -- like I said, if you put the custom_target in the subdir meson.build then you can tell meson to output it to the build directory's `iso_root/` directory and the directory will be created for you, no need to e.g. chain together `mkdir ... && cp ...`

I have tried doing as you described but it did not work,
I ended up with : 
```

kernel = executable('kernel.elf',
    src,
    include_directories : inc,
    link_args : [ '-Wl,-T' + ldscript ],
    link_depends : [ ldscript ],
)

cp = find_program('cp')
xorriso = find_program('xorriso')

kernel_iso = custom_target('limine_setup',
  input : ['limine.cfg', 'external/limine/bin/limine.sys', 'external/limine/bin/limine-cd.bin', 'external/limine/bin/limine-cd-efi.bin'],
  output : ['limine.iso'],
  command : [cp, '@INPUT@', 'iso_root'])

custom_target('kernel_iso',
  output : ['kernel.iso'],
  command : [xorriso, '-as mkisofs -b limine-cd.bin
		-no-emul-boot -boot-load-size 4 -boot-info-table
		--efi-boot limine-cd-efi.bin
		-efi-boot-part --efi-boot-image --protective-msdos-label
		iso_root -o disk.iso
		external/limine/bin/limine-deploy disk.iso'])


```
and xorriso says the following : 
```
--- stderr ---
xorriso 1.5.4 : RockRidge filesystem manipulator, libburnia project.

xorriso : FAILURE : Not a known command:  '-as mkisofs -b limine-cd.bin
                -no-emul-boot -boot-load-size 4 -boot-info-table
                --efi-boot limine-cd-efi.bin
                -efi-boot-part --efi-boot-image --protective-msdos-label
                iso_root -o $(ISO_NAME)
                external/limine/bin/limine-deploy disk.iso'

xorriso : aborting : -abort_on 'FAILURE' encountered 'FAILURE'

ninja: build stopped: subcommand failed.
```

**Reply by eli-schwartz** [reply]
> ```
> command : [xorriso, '-as mkisofs -b limine-cd.bin
> 		-no-emul-boot -boot-load-size 4 -boot-info-table
> 		--efi-boot limine-cd-efi.bin
> 		-efi-boot-part --efi-boot-image --protective-msdos-label
> 		iso_root -o disk.iso
> 		external/limine/bin/limine-deploy disk.iso'])
> ```

This appears to be passing the entire argument list for xorriso as a single string, try passing it as an array.
